src/workerd/api/streams/internal.c++

```diff
@@ -96,23 +94,23 @@ private:
   kj::Vector<kj::Array<kj::byte>> parts;
   uint64_t runningTotal = 0;
 
-  kj::Promise<PartList> loop() {
-    auto bytes = kj::heapArray<kj::byte>(4096);
+  kj::Promise<PartList> readParts() {
+    while (true) {
+      constexpr size_t bufferSize = 4096;
+      auto part = kj::heapArray<kj::byte>(bufferSize);
+      auto amount = co_await input.tryRead(part.begin(), bufferSize, bufferSize);
 
-    return input.tryRead(bytes.begin(), 1, bytes.size())
-        .then([this, bytes = kj::mv(bytes)](size_t amount) mutable
-        -> kj::Promise<PartList> {
       if (amount == 0) {
-        return KJ_MAP(p, parts) { return p.asPtr(); };
+        co_return KJ_MAP(p, parts) { return p.asPtr(); };
```

Comment:
If `PartList` were `Vector` instead of `Array`, you could avoid the `KJ_MAP` here I think? Save another allocation.